### ISSUE TITLE
training speed too slow

### ISSUE THREAD

**Issue by Google1234** [sfzhang15/RefineDet#111]
using Resnet101 , RefinneDet 512, P100 (2 gpus),batch size=6
12 hours only  2000 epoch iteration ，GPU-Util is low  ，training speed too slow 
Any suggestion？ THX
<img width="567" alt="solw" src="https://user-images.githubusercontent.com/15956648/44249148-31ec7280-a221-11e8-9c9a-03603daf3c7c.png">

**Comment by sfzhang15** [comment]
@Google1234 
We use 4 M40 GPUs (24G) to train ResNet101_RefinneDet512 with batch size = 20, i.e., 5 images per GPU, it takes about 15-20 days to finish the training.

**Comment by Google1234** [comment]
RefineDet data aug based on SSD,I guess SSD data enhancement is also time consuming
My accelerated experimental effect is here ：）

[11.pdf](https://github.com/sfzhang15/RefineDet/files/2309297/11.pdf)

[new.pdf](https://github.com/sfzhang15/RefineDet/files/2309307/new.pdf)
![1](https://user-images.githubusercontent.com/15956648/44449717-7c0a9500-a621-11e8-9219-3b32d62b9438.png)

**Comment by sfzhang15** [comment]
@Google1234 
Yes, the data augmentation in RefineDet is the same as in SSD. After seeing the files, your efficient improvement is to save the time spending on the Datum<->Mat, it is interesting. Could you share the modified files? By the way, the distort and expand operations do not need to transform the annotations.

**Comment by Google1234** [comment]
three file changed in speedup.zip
[speedup.zip](https://github.com/sfzhang15/RefineDet/files/2309528/speedup.zip)

**Comment by sfzhang15** [comment]
@Google1234 
Ok. Thanks.

**Comment by KimSoybean** [comment]
Has the training speed improved? I 've just tried you speedup.zip and feels no speed improvement. Have you changed your training code?

**Comment by Google1234** [comment]
@KimSoybean try speedup.zip replace your code

**Comment by Google1234** [comment]
@KimSoybean try use caffe time to measure time consume

**Comment by XiaYuanxiang** [comment]
@Google1234  I use speedup.zip replace my code and rebuild refinedet， but got an error, do you know how to fix it?
PROTOC src/caffe/proto/caffe.proto
CXX src/caffe/layers/euclidean_loss_layer.cpp
CXX src/caffe/layers/annotated_data_layer.cpp
CXX src/caffe/layers/lstm_unit_layer.cpp
CXX src/caffe/layers/tile_layer.cpp
CXX src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp
CXX src/caffe/layers/split_layer.cpp
CXX src/caffe/layers/batch_norm_layer.cpp
src/caffe/layers/annotated_data_layer.cpp: In member function ‘virtual void caffe::AnnotatedDataLayer<Dtype>::load_batch(caffe::Batch<Dtype>*)’:
src/caffe/layers/annotated_data_layer.cpp:259:86: error: invalid initialization of reference of type ‘const caffe::AnnotatedDatum&’ from expression of type ‘google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>’
             GenerateBatchSamples(expand_anno_groups, batch_samplers_, &sampled_bboxes);
                                                                                      ^
In file included from src/caffe/layers/annotated_data_layer.cpp:14:0:
./include/caffe/util/sampler.hpp:33:6: note: in passing argument 1 of ‘void caffe::GenerateBatchSamples(const caffe::AnnotatedDatum&, const std::vector<caffe::BatchSampler>&, std::vector<caffe::NormalizedBBox>*)’
 void GenerateBatchSamples(const AnnotatedDatum& anno_datum,
      ^
src/caffe/layers/annotated_data_layer.cpp: In instantiation of ‘void caffe::AnnotatedDataLayer<Dtype>::load_batch(caffe::Batch<Dtype>*) [with Dtype = float]’:
src/caffe/layers/annotated_data_layer.cpp:409:1:   required from here
src/caffe/layers/annotated_data_layer.cpp:212:21: error: no matching function for call to ‘caffe::DataTransformer<float>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                     this->data_transformer_->TransformAnnotation(distort_img,distort_anno_groups,do_resize, expand_bbox, do_mirror, //use img before transform
                     ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = float]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
src/caffe/layers/annotated_data_layer.cpp:243:21: error: no matching function for call to ‘caffe::DataTransformer<float>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                     this->data_transformer_->TransformAnnotation(distort_img,distort_anno_groups, do_resize, expand_bbox, do_mirror, //use img before transform
                     ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = float]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
In file included from ./include/caffe/common.hpp:6:0,
                 from ./include/caffe/blob.hpp:8,
                 from ./include/caffe/data_transformer.hpp:8,
                 from src/caffe/layers/annotated_data_layer.cpp:11:
src/caffe/layers/annotated_data_layer.cpp:251:29: warning: the address of ‘expand_img’ will never be NULL [-Waddress]
         CHECK((&expand_img) != NULL);
                             ^
src/caffe/layers/annotated_data_layer.cpp:252:37: warning: the address of ‘expand_anno_groups’ will never be NULL [-Waddress]
         CHECK((&expand_anno_groups) != NULL);
                                     ^
src/caffe/layers/annotated_data_layer.cpp:270:17: error: no matching function for call to ‘caffe::DataTransformer<float>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                 this->data_transformer_->TransformAnnotation(expand_img,expand_anno_groups, do_resize, crop_bbox, do_mirror, //use img before crop
                 ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = float]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
In file included from ./include/caffe/common.hpp:6:0,
                 from ./include/caffe/blob.hpp:8,
                 from ./include/caffe/data_transformer.hpp:8,
                 from src/caffe/layers/annotated_data_layer.cpp:11:
src/caffe/layers/annotated_data_layer.cpp:281:29: warning: the address of ‘sample_img’ will never be NULL [-Waddress]
         CHECK((&sample_img) != NULL);
                             ^
src/caffe/layers/annotated_data_layer.cpp:282:37: warning: the address of ‘sample_anno_groups’ will never be NULL [-Waddress]
         CHECK((&sample_anno_groups) != NULL);
                                     ^
src/caffe/layers/annotated_data_layer.cpp:324:17: error: no matching function for call to ‘caffe::DataTransformer<float>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                 this->data_transformer_->TransformAnnotation(sample_img,sample_anno_groups,do_resize,crop_bbox,do_mirror,//use img before Transform
                 ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = float]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
src/caffe/layers/annotated_data_layer.cpp: In instantiation of ‘void caffe::AnnotatedDataLayer<Dtype>::load_batch(caffe::Batch<Dtype>*) [with Dtype = double]’:
src/caffe/layers/annotated_data_layer.cpp:409:1:   required from here
src/caffe/layers/annotated_data_layer.cpp:212:21: error: no matching function for call to ‘caffe::DataTransformer<double>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                     this->data_transformer_->TransformAnnotation(distort_img,distort_anno_groups,do_resize, expand_bbox, do_mirror, //use img before transform
                     ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = double]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
src/caffe/layers/annotated_data_layer.cpp:243:21: error: no matching function for call to ‘caffe::DataTransformer<double>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                     this->data_transformer_->TransformAnnotation(distort_img,distort_anno_groups, do_resize, expand_bbox, do_mirror, //use img before transform
                     ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = double]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
In file included from ./include/caffe/common.hpp:6:0,
                 from ./include/caffe/blob.hpp:8,
                 from ./include/caffe/data_transformer.hpp:8,
                 from src/caffe/layers/annotated_data_layer.cpp:11:
src/caffe/layers/annotated_data_layer.cpp:251:29: warning: the address of ‘expand_img’ will never be NULL [-Waddress]
         CHECK((&expand_img) != NULL);
                             ^
src/caffe/layers/annotated_data_layer.cpp:252:37: warning: the address of ‘expand_anno_groups’ will never be NULL [-Waddress]
         CHECK((&expand_anno_groups) != NULL);
                                     ^
src/caffe/layers/annotated_data_layer.cpp:270:17: error: no matching function for call to ‘caffe::DataTransformer<double>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, const bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                 this->data_transformer_->TransformAnnotation(expand_img,expand_anno_groups, do_resize, crop_bbox, do_mirror, //use img before crop
                 ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = double]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
In file included from ./include/caffe/common.hpp:6:0,
                 from ./include/caffe/blob.hpp:8,
                 from ./include/caffe/data_transformer.hpp:8,
                 from src/caffe/layers/annotated_data_layer.cpp:11:
src/caffe/layers/annotated_data_layer.cpp:281:29: warning: the address of ‘sample_img’ will never be NULL [-Waddress]
         CHECK((&sample_img) != NULL);
                             ^
src/caffe/layers/annotated_data_layer.cpp:282:37: warning: the address of ‘sample_anno_groups’ will never be NULL [-Waddress]
         CHECK((&sample_anno_groups) != NULL);
                                     ^
src/caffe/layers/annotated_data_layer.cpp:324:17: error: no matching function for call to ‘caffe::DataTransformer<double>::TransformAnnotation(cv::Mat&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>&, const bool&, caffe::NormalizedBBox&, bool&, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*)’
                 this->data_transformer_->TransformAnnotation(sample_img,sample_anno_groups,do_resize,crop_bbox,do_mirror,//use img before Transform
                 ^
In file included from src/caffe/layers/annotated_data_layer.cpp:11:0:
./include/caffe/data_transformer.hpp:99:8: note: candidate: void caffe::DataTransformer<Dtype>::TransformAnnotation(const caffe::AnnotatedDatum&, bool, const caffe::NormalizedBBox&, bool, google::protobuf::RepeatedPtrField<caffe::AnnotationGroup>*) [with Dtype = double]
   void TransformAnnotation(
        ^
./include/caffe/data_transformer.hpp:99:8: note:   candidate expects 5 arguments, 6 provided
Makefile:575: recipe for target '.build_release/src/caffe/layers/annotated_data_layer.o' failed
make: *** [.build_release/src/caffe/layers/annotated_data_layer.o] Error 1
make: *** 正在等待未完成的任务....

**Comment by Google1234** [comment]
check this :
./include/caffe/data_transformer.hpp:99:8: note: candidate expects 5 arguments, 6 provided

**Comment by XiaYuanxiang** [comment]
@Google1234 
I check data_transformer.hpp and data_transformer.cpp, there are two overload function as follows, 
which have 5 and 6 arguments,  i can't figure out what's wrong with that

 void TransformAnnotation(
      const AnnotatedDatum& anno_datum, const bool do_resize,
      const NormalizedBBox& crop_bbox, const bool do_mirror,
      RepeatedPtrField<AnnotationGroup>* transformed_anno_group_all);
  void TransformAnnotation(
      const cv::Mat& img, const RepeatedPtrField<AnnotationGroup>& src_anno_group,
      const bool do_resize, const NormalizedBBox& crop_bbox, const bool do_mirror,
      RepeatedPtrField<AnnotationGroup>* transformed_anno_group_all);
